### PR TITLE
Fetch reward account balance when reading or listing wallets

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#904 

# Overview

- [x] Byron wallets no longer have a `reward`-field in their `balance`.
- [x] Fetch reward balance when reading or listing Shelley wallets:
    - [x] I have added a `HasRewardAccount` an instance for `SeqState`.
    - [x] I added `getAccountBalance` to `NetworkLayer` which calls `getAccountState` of `JormungandrClient`
    - [x] I added `Cardano.Wallet.fetchRewardBalance` calling `getAccountBalance` with some error-handling.
    - [x] I made sure created Api Wallets ask `fetchRewardBalance` for the reward balance.
- [x] I required the controlled-stake of the pool to increase in `STAKE_POOL_JOIN_01`
- [x] Commented integration test checking that the `rewardBalance` increases

# Comments

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: <s>This doesn't compile when I require `HasNetworkLayer t ctx` from `W.fetchWalletRewardBalance `</s> 🤔 

Ok, explicit type application helped.

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: **[suggestion]**

Drop `Wallet` from the name

**Why**

There's no much ambiguity about where the reward balance comes from. It makes the overall clearer.

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: **[change required]**

Fix the top-level signature by removing `s ~ SeqState n k` and using `hasRewardAccount s` instead

**Why**

As the `FIXME` mentions, this was an ugly hack that would better be tackled by using a better abstraction.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: **[question]**

Why a plural form here :grimacing:  ?

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: **[change required]**

Replace `error "TODO"` with `Nothing`

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: **[change required]**

Do not silently swallow an error. If one error deserves to be ignore, use a clear pattern match to show it but avoid holes. 

**Why**

There are two reasons, the first one is:

- https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#prefer-pattern-matching-to-equality-testing-on-sum-types

And the second is that, if the account doesn't exist on the blockchain, we may want to default to a value of `0`. However, in case we get any sort of error, we should either fail, or, at the very least, log an error.
Note also that this is also an abuse of the `liftE2` function IMO which does much more than just lifting the error. I'd suggest moving the default handling logic to another function and to compose.

File: lib/core-integration/src/Test/Integration/Framework/DSL.hs
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Framework/DSL.hs]: **[sugestion]**

You may want to rebase on latest master, this lens now exists as `balanceReward`.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: **[suggestion]**

Remove the extra `do` and newlines.

**Why**

It creates needless noise in the review, as well as cluttering a bit the function itself. Unless there's a good reason to do so?

File: lib/core/src/Cardano/Wallet.hs
Comment (Anviking) [reply lib/core/src/Cardano/Wallet.hs]: `rewardAccount` returns a `Maybe ChimericAccount`, and we need a `k 'AddressK XPub` here.

We could change `rewardAccount` to be `Maybe (k 'AddressK XPub)`, but we'd need to involve `k` in the `HasRewardAccount s` class… 🤔

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: I'm not sure about the `Has`-prefix. Seeing the constraint `HasRewardAccount s` I would like to believe that `s` **has** a reward account. But it doesn't mean that! `rewardAccount @s` could return `Nothing`!

Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: Why could it though ? I don't think `HasRewardAccount` should be defined on `RndState` in the first place actually.

Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: That's kinda the point of all these instances and the way the wallet engine works. You only define the capabilities on the relevant concrete types. 

`RndState` doesn't implement `DelegationAddress` for instance, and that's very fine :)

Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: > I don't think HasRewardAccount should be defined on RndState in the first place actually.

No, I agree. (conceptually)

It simplified things in `Api.Server`. I will see now if I can do anything.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: Without `balance` as suffix, I thought `rewards` sounded better than `reward`. (@jonathanknowles ?)

`rewardBal[ance]` might be a safe bet though.

File: lib/core/src/Cardano/Wallet.hs
Comment (Anviking) [inline lib/core/src/Cardano/Wallet.hs:1012]: This is a private key. Maybe we should rename `HasRewardAccount.rewardAccount` to `rewardAccountXPub` or something.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (Anviking) [inline lib/core/src/Cardano/Wallet/Api/Server.hs:1244]: Not sure if we have to update swagger.yaml.

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs:695]: :+1:

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet.hs]: :+1:

File: lib/core/src/Cardano/Wallet.hs
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet.hs:1012]: `rewardAccountXPub` is redundant with the type signature. Instead, we could make `deriveRewardAccount` part of the `HasRewardInterface` to make sure they're done consistently. Pretty minor here however.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/Api/Server.hs]: **[suggestion]**

rewrite to `mapM (getWallet ctx) (ApiT<$> wids)` 

**Why**

Avoid logic duplication in `mkWallet`

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (KtorZ) [reply lib/core/src/Cardano/Wallet/Api/Server.hs:1244]: It does need to be updated yes: https://input-output-hk.github.io/cardano-wallet/api/edge/#operation/getByronWallet

File: lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs
Comment (KtorZ) [inline lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs:275]: :+1:

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs
Comment (KtorZ) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs]: **[change required]** 

Use `faucetUtxoAmt` (from Test.Integration.Framework.DSL) and `stakedelegationfee` (from Test.Integration.Framework.TestData)

**Why**

https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#prefer-named-constants-over-magic-numbers

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs
Comment (KtorZ) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:211]: **[change required]**

Move this expectation into a separate scenario

**Why**

This is yet a different scenarios that would deserve its own test for clarity and debugging. Try to keep scenarios focused on doing one thing.

**Review by KtorZ — CHANGES_REQUESTED** [review]

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs
Comment (Anviking) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/StakePools.hs:211]: Done. We should perhaps be mindful about that these tests take quite a lot of time though.

File: lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Random.hs]: Changed

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Api/Server.hs:1244]: f8c63cc

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: Unfortunately doesn't work since `listWallets` needs to sort on the creation time in `(ApiWallet, UTCTime)`, while `getWallet` only can return `ApiWallet`.

Maybe we can rewrite some of the `getWalletWithCreationTime` and `mkWallet`-indirection, however…

Comment (Anviking) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: 06b04a5

I removed `mkApiWallet` and `mkApiByronWallet` which were arguments to `getWalletWithCreationTime`.

I merged `getWalletWithCreationTime` with `mkApiWallet` and added `getByronWalletWithCreationTime`.

This way there is less indirection.

**Review by KtorZ — COMMENTED** [review]
LGTM. I'll look into enabling the remaining integration tests on top of rc4 before merging.

File: lib/core/src/Cardano/Wallet/Api/Server.hs
Comment (jonathanknowles) [reply lib/core/src/Cardano/Wallet/Api/Server.hs]: Either "reward" or "rewards" are fine, I think!

However, perhaps "rewards" has the unwanted connotation of implying a list or some other container.

`rewardBalance` is also a good compromise.

**Comment by KtorZ** [comment]
Okay, this PR passes with Jörmungandr `rc7`. Rewards do accumulate (though they still behave a bit strange IMO.. question asked, waiting for an answer..)

Will merge after #1117

**Review by KtorZ — APPROVED** [review]

**Pull request merged**
